File: js_solutions/valid-parentheses.js
function isValid(string) {
  const stack = [];
  const matchingPairs = { "(": ")", "[": "]", "{": "}" };
  for (char of string) {
    if (Object.keys(matchingPairs).includes(char)) {
      stack.push(char);
    } else {
      const lastOpen = stack.pop();
      if (matchingPairs[lastOpen] !== char) return false;
    }
  }
  return stack.length === 0;
}

File: ruby_solutions/easy/valid_parentheses.rb
# frozen_string_literal: true

def valid_parentheses(string)
  stack = []
  matching_pairs = { "(" => ")", "[" => "]", "{" => "}" }
  string.each_char do |char|
    if matching_pairs.key?(char)
      stack.push(char)
    else
      opening_paren = stack.pop
      return false unless matching_pairs[opening_paren] == char
    end
  end
  stack.empty?
end
